pilot/platform/kube/inject/inject.go

```diff
@@ -322,172 +330,51 @@ func timeString(dur *duration.Duration) string {
+// in string form for proper template interpolation
+type ProxySettings struct {
+	DrainDuration          string
+	ParentShutdownDuration string
+	DiscoveryRefreshDelay  string
+	ConnectTimeout         string
+	ControlPlaneAuthPolicy string
+}
 
-	// sidecar proxy container
-	args := []string{"proxy", "sidecar"}
+func injectIntoSpec(p *Params, spec *v1.PodSpec, metadata *metav1.ObjectMeta) {
 
-	if p.Verbosity > 0 {
-		args = append(args, "-v", strconv.Itoa(p.Verbosity))
-	}
+	pSettings := ProxySettings{timeString(p.Mesh.DefaultConfig.DrainDuration),
+		timeString(p.Mesh.DefaultConfig.ParentShutdownDuration),
+		timeString(p.Mesh.DefaultConfig.DiscoveryRefreshDelay),
+		timeString(p.Mesh.DefaultConfig.ConnectTimeout), p.Mesh.DefaultConfig.ControlPlaneAuthPolicy.String()}
 
-	serviceCluster := p.Mesh.DefaultConfig.ServiceCluster
+	data := struct {
+		Spec           *v1.PodSpec
+		ServiceCluster string
+		MConfig        *Params
+		PConfig        *ProxySettings
+	}{spec, "", p, &pSettings}
 
+	data.ServiceCluster = p.Mesh.DefaultConfig.ServiceCluster
 	// If 'app' label is available, use it as the default service cluster
 	if val, ok := metadata.GetLabels()["app"]; ok {
-		serviceCluster = val
+		data.ServiceCluster = val
 	}
 
-	// set all proxy config flags
-	args = append(args, "--configPath", p.Mesh.DefaultConfig.ConfigPath)
-	args = append(args, "--binaryPath", p.Mesh.DefaultConfig.BinaryPath)
-	args = append(args, "--serviceCluster", serviceCluster)
-	args = append(args, "--drainDuration", timeString(p.Mesh.DefaultConfig.DrainDuration))
-	args = append(args, "--parentShutdownDuration", timeString(p.Mesh.DefaultConfig.ParentShutdownDuration))
-	args = append(args, "--discoveryAddress", p.Mesh.DefaultConfig.DiscoveryAddress)
-	args = append(args, "--discoveryRefreshDelay", timeString(p.Mesh.DefaultConfig.DiscoveryRefreshDelay))
-	args = append(args, "--zipkinAddress", p.Mesh.DefaultConfig.ZipkinAddress)
-	args = append(args, "--connectTimeout", timeString(p.Mesh.DefaultConfig.ConnectTimeout))
-	args = append(args, "--statsdUdpAddress", p.Mesh.DefaultConfig.StatsdUdpAddress)
-	args = append(args, "--proxyAdminPort", fmt.Sprintf("%d", p.Mesh.DefaultConfig.ProxyAdminPort))
-	args = append(args, "--controlPlaneAuthPolicy", p.Mesh.DefaultConfig.ControlPlaneAuthPolicy.String())
-
-	volumeMounts := []v1.VolumeMount{
-		{
-			Name:      istioEnvoyConfigVolumeName,
-			MountPath: p.Mesh.DefaultConfig.ConfigPath,
-		},
-	}
+	var tmpl bytes.Buffer
 
-	spec.Volumes = append(spec.Volumes,
-		v1.Volume{
-			Name: istioEnvoyConfigVolumeName,
-			VolumeSource: v1.VolumeSource{
-				EmptyDir: &v1.EmptyDirVolumeSource{
-					Medium: v1.StorageMediumMemory,
-				},
-			},
-		})
-
-	volumeMounts = append(volumeMounts, v1.VolumeMount{
-		Name:      istioCertVolumeName,
-		ReadOnly:  true,
-		MountPath: proxy.AuthCertsPath,
-	})
-
-	sa := spec.ServiceAccountName
-	if sa == "" {
-		sa = "default"
+	t := template.Must(template.New("inject").Parse(productionTemplate))
```

Comment:
We'll want to eventually load this from a ConfigMap to allow for user customization.

Reply:
Right, I also suggest adding a descriptor to the template (e.g. that it is go-template). We might want to use other templating engines down the road (like this one https://github.com/kyessenov/envoymesh/blob/master/inject.jsonnet).

Reply:
Can I put a comment to allow for configuration before intializer/webhook goes GA?